[CDN] Fixes #12152: Add custom domain BYOC support.

lsmith130 opened the pull request:
Adds CDN custom domain BYOC support and fixes CDN-managed certs for Verizon and Microsoft SKUs. Fixes #12152 and #9894.

**History Notes:**  

[CDN] az cdn custom-domain enable-https: Add BYOC support.  
[CDN] az cdn custom-domain enable-https: Fix enabling custom HTTPS with CDN managed certificates for Standard_Verizon and Standard_Microsoft SKUs.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

yonzhan commented:
add to S167

lsmith130 commented:
Hi again @Juliehzl, what's the timeframe for when I can get a review for this?

Juliehzl commented on src/azure-cli/azure/cli/command_modules/cdn/_params.py:
```diff
@@ -135,6 +135,21 @@ def load_arguments(self, _):
+        c.argument('min_tls_version',
+                   help='The minimum TLS version required for the custom domain.',
+                   arg_type=get_enum_type(['None', '1.0', '1.2']))
+        c.argument('user_cert_protocol_type',
+                   help='The protocol type of the certificate. Options are "sni" and "ip". Defaults to "sni".')
+        c.argument('user_cert_subscription_id', help='The subscription id of the KeyVault certificate')
+        c.argument('user_cert_group_name', help='The resource group of the KeyVault certificate')
+        c.argument('user_cert_vault_name', help='The vault name of the KeyVault certificate')
+        c.argument('user_cert_secret_name', help='The secret name of the KeyVault certificate')
+        c.argument('user_cert_secret_version', help='The secret version of the KeyVault certificate')
```
Can we group these arguments using `arg_group` kwargs?

lsmith130 replied:
Done.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/cdn/custom.py:
```diff
@@ -492,6 +492,101 @@ def create_custom_domain(client, resource_group_name, profile_name, endpoint_nam
+                                   user_cert_protocol_type]):
+
+        # If any BYOC flags are set, make sure they all are.
+        if any(x is not None for x in [user_cert_subscription_id,
```
It looks the condition should be `x is None` to check if there is an element lost, right?

lsmith130 replied:
Fixed, good catch.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/cdn/_params.py:
```diff
@@ -135,6 +135,21 @@ def load_arguments(self, _):
+        c.argument('min_tls_version',
+                   help='The minimum TLS version required for the custom domain.',
+                   arg_type=get_enum_type(['None', '1.0', '1.2']))
+        c.argument('user_cert_protocol_type',
```
So for user_cert_protocol_type, it should be enum type?

lsmith130 replied:
Fixed.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/cdn/tests/latest/recordings/test_cdn_waf_policy_crud.yaml:
```diff
@@ -6718,65 +4063,6 @@ interactions:
     status:
       code: 200
       message: OK
-- request:
```
Why are there so many recording deleted?

lsmith130 replied:
I believe this is due to some change in the cli core, it seems like the authentication request at the beginning of some tests is no longer needed.

haroldrandom commented on src/azure-cli/azure/cli/command_modules/cdn/_help.py:
```diff
@@ -66,6 +66,20 @@
+examples:
+  - name: Enable HTTPS using a CDN-managed certificate
+    text: >
+        az cdn custom-domain enable-https -g group --profile-name profile --endpoint-name endpoint \\
```
You can remove `\\` to let CLI folder it automatically since you are using `>` here to indicate YAML block. Othersize the newline wrapper `\\` might behave unexpected based on different terminal windows size.

lsmith130 replied:
Fixed.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/cdn/_help.py:
```diff
@@ -66,6 +66,20 @@
+  - name: Enable HTTPS using a CDN-managed certificate and a minimum TLS version of 1.2
+    text: >
+        az cdn custom-domain enable-https -g group --profile-name profile --endpoint-name endpoint \\
+            -n domain-name --min-tls-version=1.2
```
Can we use --min-tls-version 1.2 to keep consistent with other arguments?

lsmith130 replied:
Fixed.

yonzhan commented:
add to S169

MyronFanQiu commented:
@lsmith130 Could you address the comments so that we can merge this PR to fix issues

MyronFanQiu commented:
/azp run Azure.azure-cli

lsmith130 commented:
@MyronFanQiu, I've addressed the comments and CI failures.

MyronFanQiu commented on src/azure-cli/azure/cli/command_modules/cdn/_params.py:
```diff
@@ -135,6 +135,33 @@ def load_arguments(self, _):
+                   arg_group='Bring Your Own Certificate',
+                   help='The protocol type of the certificate.',
+                   arg_type=get_enum_type(['sni', 'ip']))
+        c.argument('user_cert_subscription_id',
```
is it possible to support one secret id argument?

lsmith130 replied:
This is a non-trivial request, as we would need to add extra logic to retrieve the correct resource group name for the secret, which is not included in the secret id because it doesn't use a standard Azure resource id. We would also have to correctly handle permissions issues where the user may not have permission to perform that lookup even though they do have permission to reference the key through CDN. Can we include this later as a separate PR if customers request it?

yonzhan commented:
add to S170

Wyren commented on src/azure-cli/azure/cli/command_modules/cdn/tests/latest/recordings/test_cdn_custom_domain_errors.yaml:
```diff
@@ -0,0 +1,892 @@
+      accept-language:
+      - en-US
+    method: GET
+    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/cli_test_cdn_domain000001/providers/Microsoft.Cdn/profiles/cdnprofile1?api-version=2019-06-15-preview
```
This is a get method and a call for profile. So this doesn't match the command you're describing.

lsmith130 replied:
This profile GET is only part of the setup for the test. If you look below, later requests are actually performing enableCustomHttps POST commands on the custom domains

Wyren commented on src/azure-cli/azure/cli/command_modules/cdn/tests/latest/recordings/test_cdn_custom_domain_https_msft.yaml:
```diff
@@ -0,0 +1,2323 @@
+      - en-US
+    method: GET
+    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Cdn/profiles/profile123?api-version=2019-06-15-preview
+  response:
```
Same here. This is for a pull request to fix this call and the tests seem off. It's weird.

lsmith130 replied:
Again, the tests must perform required setup before testing the actual area of interest.

Wyren replied:
Thanks for the explanation. For some reason I assumed these were combined with the other tests considering the setup could be handled in that manor. But this makes it a lot clearer.

My apologies for the comments going a bit to far. The lock-down has put me on edge and frustration got the better of me.

Wyren commented on src/azure-cli/azure/cli/command_modules/cdn/tests/latest/recordings/test_cdn_custom_domain_https_msft.yaml:
```diff
@@ -0,0 +1,2323 @@
+      accept-language:
+      - en-US
+    method: GET
+    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Cdn/profiles/profile123/endpoints/cdn-cli-test-4/customDomains/customdomain000002?api-version=2019-06-15-preview
```
Why is this happening in multiple places...

lsmith130 replied:
This GET request is retrieving the custom domain to verify its status is now correct after the previous POST enableCustomHttps request.

Wyren commented on src/azure-cli/azure/cli/command_modules/cdn/tests/latest/recordings/test_cdn_custom_domain_https_msft.yaml:
```diff
@@ -0,0 +1,2323 @@
+      accept-language:
+      - en-US
+    method: GET
+    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Cdn/profiles/profile123?api-version=2019-06-15-preview
```
Am I maybe understanding these test wrong or what's the deal.

Wyren commented on src/azure-cli/azure/cli/command_modules/cdn/tests/latest/recordings/test_cdn_custom_domain_https_msft.yaml:
```diff
@@ -0,0 +1,2323 @@
+      accept-language:
+      - en-US
+    method: GET
+    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest.rg000001/providers/Microsoft.Cdn/profiles/profile123/endpoints/cdn-cli-test-4/customDomains/customdomain000003?api-version=2019-06-15-preview
```
Do I need to put this at every enable-https request that has a get method? Because some do use the proper uri en method. So I don't know if it really is a mistake anymore. I think the message is getting through right?

lsmith130 replied:
Same as above. This is a required part of the test. Since these are not unit tests it is impossible to test only the functionality of interest without making addition setup calls and reading state to verify correct behavior. Please let me know if I have not addressed your concerns in my comments.

lsmith130 commented:
@Juliehzl is it possible to approve this PR? I've addressed all questions and requested changes.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/cdn/_params.py:
```diff
@@ -135,6 +135,33 @@ def load_arguments(self, _):
         c.argument('location', validator=get_default_location_from_resource_group)
 
+    with self.argument_context('cdn custom-domain enable-https') as c:
+        c.argument('profile_name', id_part=None, help='Name of the parent profile.')
```
Prefer ['--name', '-n’ ]e

lsmith130 replied:
The name_arg_type parameter to the argument custom_domain_name already makes the parameter use --name, -n as requested. See the example for `cdn custom-domain enable-https`

yonzhan commented:
add to S171

lsmith130 commented:
@haroldrandom could you take another look at this PR? I've addressed all comments and I believe it just needs code-owner approval before it can be merged.
